inginious/frontend/pages/course_admin/student_list.py

```diff
@@ -192,35 +198,92 @@ def post_audiences(self, course, data, active_tab, msg, error):
-
-                for index, new_audience in enumerate(audiences):
-                    # In case of file upload, no id specified
-                    new_audience['_id'] = new_audience['_id'] if '_id' in new_audience else 'None'
-
-                    # Update the audience
-                    audience, errors = self.update_audience(course, new_audience['_id'], new_audience)
-
-                    # If file upload was done, get the default audience id
-                    audienceid = audience['_id']
-                    errored_students += errors
+            if "audiencefile" in data and 'upload_audiences_creation' in data:
+                # get the Werkzeug datastructures.FileStorage object.
+                # The stream of this object is the stream body of the uploaded file.
+                # Furthermore, FileStorage.stream seems to inherit ʻio.BufferedIOBase`, so this stream should be boiled.
+                # As reader return an iterator and that we iterate twice, it is faster to cast into a list.
+                csv_data = list(csv.reader(io.TextIOWrapper(data["audiencefile"], encoding='utf-8')))
+                # Define used variables.
+                students_per_audience = {}
+                tutors_per_audience = {}
+                course_students = []
+                course_tutors = []
+                audiences = []
+                # Check correctness of CSV structure.
+                for line in csv_data:
+                    if len(line) != 4:
+                        msg["audiences"] = _("File wrongly formatted.")
+                        error["audiences"] = True
+                if "audiences" not in error or not error["audiences"]:
+                    stud_list, aud_li, oth_stu, u_info = self.get_user_lists(course)
+                    courseid = course.get_id()
+                    # Fully remove previous audiences.
+                    self.database.audiences.delete_many({"courseid": courseid})
+                    # read datas from CSV.
+                    for user_id, field, role, description in csv_data:
+                        user_id = user_id.strip()
+                        field = field.strip()
+                        role = role.strip()
+                        if description != "":
+                            description = description.strip()
+                        if field not in ["username", "email", "realname"] + \
+                                list(self.user_manager.get_auth_methods().keys()):
+                            msg["audiences"] = _("Field was not recognized: ") + field
+                            error["audiences"] = True
+                            continue
+                        if role not in ["student", "tutor"]:
+                            msg["audiences"] = _("Unknown role: ") + role
+                            error["audiences"] = True
+                            continue
+                        if field != "username":
```

Comment:
Why ? A non-existing username should not be accepted.

Reply:
It was to avoid unecessary request to database as I work with username after but u re right that a non-existing user is actually passing through.